File: example-dual-use/print-from-creds.js
#!/usr/bin/env node

// Example 2: Print credentials from .creds file
// This demonstrates the runtime approach where credentials are loaded from files

const crypto = require('crypto');
const fs = require('fs');
const path = require('path');

// Load the encryption key
const keyPath = path.join('.credential-code', 'encryption-key.txt');
const keyBase64 = fs.readFileSync(keyPath, 'utf8').trim();
const key = Buffer.from(keyBase64, 'base64');

// Load the .creds file
const credsPath = path.join('Generated', 'credentials.creds');
const credsData = JSON.parse(fs.readFileSync(credsPath, 'utf8'));

// Function to decrypt a credential
function decrypt(credentialKey) {
    const encrypted = credsData.credentials[credentialKey];
    if (!encrypted) {
        throw new Error(`Credential ${credentialKey} not found`);
    }
    
    // Decode from base64
    const data = Buffer.from(encrypted.data, 'base64');
    const nonce = Buffer.from(encrypted.nonce, 'base64');
    const tag = Buffer.from(encrypted.tag, 'base64');
    
    // Create decipher
    const decipher = crypto.createDecipheriv('aes-256-gcm', key, nonce);
    decipher.setAuthTag(tag);
    
    // Decrypt
    const decrypted = Buffer.concat([
        decipher.update(data),
        decipher.final()
    ]);
    
    return decrypted.toString('utf8');
}

// Print all credentials
console.log('=== Credentials from .creds file ===');
const result = {};

try {
    // Get all credential keys
    const keys = Object.keys(credsData.credentials);
    
    // Decrypt each one
    for (const key of keys) {
        result[key] = decrypt(key);
    }
    
    // Print as JSON
    console.log(JSON.stringify(result, null, 2));
    
} catch (error) {
    console.error('Error:', error.message);
    console.error('Make sure you have generated credentials with: credential-code generate');
}

File: example-dual-use/print-from-creds.py
#!/usr/bin/env python3

# Example 2b: Print credentials from .creds file (Python version)
# This demonstrates the runtime approach in Python
#
# NOTE: This example shows manual decryption for educational purposes.
# For production use, see the utility library at /utilities/python/

import json
import base64
from pathlib import Path
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.backends import default_backend

def decrypt_credential(key_bytes, encrypted_data):
    """Decrypt a single credential using AES-256-GCM"""
    # Decode components
    ciphertext = base64.b64decode(encrypted_data['data'])
    nonce = base64.b64decode(encrypted_data['nonce'])
    tag = base64.b64decode(encrypted_data['tag'])
    
    # Create cipher
    cipher = Cipher(
        algorithms.AES(key_bytes),
        modes.GCM(nonce, tag),
        backend=default_backend()
    )
    
    # Decrypt
    decryptor = cipher.decryptor()
    plaintext = decryptor.update(ciphertext) + decryptor.finalize()
    
    return plaintext.decode('utf-8')

def main():
    try:
        # Load encryption key
        key_path = Path('.credential-code/encryption-key.txt')
        key_base64 = key_path.read_text().strip()
        key_bytes = base64.b64decode(key_base64)
        
        # Load .creds file
        creds_path = Path('Generated/credentials.creds')
        creds_data = json.loads(creds_path.read_text())
        
        print('=== Credentials from .creds file (Python) ===')
        
        # Decrypt all credentials
        result = {}
        for cred_name, encrypted_data in creds_data['credentials'].items():
            result[cred_name] = decrypt_credential(key_bytes, encrypted_data)
        
        # Print as JSON
        print(json.dumps(result, indent=2))
        
    except FileNotFoundError as e:
        print(f"Error: File not found - {e}")
        print("Make sure you have generated credentials with: credential-code generate")
    except Exception as e:
        print(f"Error: {e}")

if __name__ == '__main__':
    main()